tailnet/conn.go

```diff
@@ -395,9 +399,56 @@ func (c *Conn) Status() *ipnstate.Status {
 	return sb.Status()
 }
 
-// Ping sends a ping to the Wireguard engine.
-func (c *Conn) Ping(ip netip.Addr, pingType tailcfg.PingType, cb func(*ipnstate.PingResult)) {
-	c.wireguardEngine.Ping(ip, pingType, cb)
+// Ping sends a Disco ping to the Wireguard engine.
+func (c *Conn) Ping(ctx context.Context, ip netip.Addr) (time.Duration, error) {
+	errCh := make(chan error, 1)
+	durCh := make(chan time.Duration, 1)
+	go c.wireguardEngine.Ping(ip, tailcfg.PingDisco, func(pr *ipnstate.PingResult) {
+		if pr.Err != "" {
+			errCh <- xerrors.New(pr.Err)
+			return
+		}
+		durCh <- time.Duration(pr.LatencySeconds * float64(time.Second))
+	})
+	select {
+	case err := <-errCh:
+		return 0, err
+	case <-ctx.Done():
+		return 0, ctx.Err()
+	case dur := <-durCh:
+		return dur, nil
+	}
+}
+
+// AwaitReachable pings the provided IP continually until the
+// address is reachable. It's the callers responsibility to provide
+// a timeout, otherwise this function will block forever.
+func (c *Conn) AwaitReachable(ctx context.Context, ip netip.Addr) bool {
+	ticker := time.NewTicker(time.Millisecond * 100)
+	defer ticker.Stop()
+	completedCtx, completed := context.WithCancel(ctx)
+	run := func() {
+		ctx, cancelFunc := context.WithTimeout(completedCtx, time.Second)
```

Comment:
@kylecarbs Couldn't this timeout result in no ping ever completing if RTT is >1s?

Also wondering if there's a potential edge-case race in `completedCtx` being based on `ctx`. Let's say caller does `cancel` on their context, and ultimately in the select below `completedCtx.Done()` is selected to resolve before `ctx.Done()`.

Reply:
@mafredri you are certainly correct. I won't have time to tackle this today, but I'd appreciate if you went through this code and cleaned up any races you might encounter. You have a much better eye for these problems than I!